add fault-tolerance and recovery (#246)
tikv/website: content/docs/5.1/concepts/explore-tikv-features/fault-tolerance.md

---
title: Fault tolerance and recovery
description: Fault tolerance, recovery
menu:
    "5.1":
        parent: Features
        weight: 2
---

> This page walks you through a simple demonstration of how TiKV remains available during and recovers after, failure.

Author: Seems they are the same meaning.

> 1. Starting with a 6-node local cluster with the default 3-way replication.

Author: Merge in the next batch.

2. Clone and compile tool [go-ycsb](https://github.com/pingcap/go-ycsb) in local for launch a workload
3. Install [client-py](https://github.com/tikv/client-py) to interact with the TiKV cluster.

## Step 1. Start a 6-node cluster

With a new terminal session, use `tiup playground` command to launch a 6-node local cluster.
```sh
tiup playground --mode tikv-slim --kv 6
```
This command will give you a hint about components' addresses. It will be used in the following steps:
```txt
Playground Bootstrapping...
Start pd instance
Start tikv instance
Start tikv instance
Start tikv instance
Start tikv instance
Start tikv instance
Start tikv instance
PD client endpoints: [127.0.0.1:2379]
To view the Prometheus: http://127.0.0.1:44549
To view the Grafana: http://127.0.0.1:3000
```

**By now, each region in this cluster will have 3 replicas according to the default configuration.**

## Step 2. Write data

On another terminal session, use [go-ycsb](https://github.com/pingcap/go-ycsb) to launch a workload.

1. Clone the `go-ycsb` from GitHub.
    ```sh
    git clone https://github.com/pingcap/go-ycsb.git
    ```
2. Build the application from the source.
    ```sh
    make

>     ```

Reviewer: I suggest to remove from `Prerequisites`.

>     ```

Author: Make sense.

3. Load a workload using `go-ycsb` with **10000** keys.
    ```sh
    ./bin/go-ycsb load tikv -P workloads/workloada -p tikv.pd="127.0.0.1:2379" -p tikv.type="raw" -p recordcount=1000000
    ```
    This command will output the following content:

>     ```txt

Reviewer: leave an empty line between normal text and code

Author: Noted.

## Step 4. Run a sample workload.

Go to the source directory of `go-ycsb`, use the following command to run the `workloada` from YCSB benchmark, simulating multiple client connections, each performing mixed 50% read and 50% write operations.

```sh
./bin/go-ycsb run tikv -P workloads/workloada -p tikv.pd="127.0.0.1:2379" -p tikv.type="raw" -p tikv.conncount=16 -p threadcount=16 -p recordcount=10000 -p operationcount=1000000
```

you'll see per-operation statistics print to standard output every second.

```txt
...
READ   - Takes(s): 10.0, Count: 7948, OPS: 796.2, Avg(us): 395, Min(us): 72, Max(us): 20545, 99th(us): 2000, 99.9th(us): 19000, 99.99th(us): 21000
UPDATE - Takes(s): 10.0, Count: 7945, OPS: 796.8, Avg(us): 19690, Min(us): 11589, Max(us): 40429, 99th(us): 34000, 99.9th(us): 41000, 99.99th(us): 41000
READ   - Takes(s): 20.0, Count: 15858, OPS: 793.6, Avg(us): 380, Min(us): 68, Max(us): 20545, 99th(us): 2000, 99.9th(us): 3000, 99.99th(us): 21000
UPDATE - Takes(s): 20.0, Count: 15799, OPS: 791.1, Avg(us): 19834, Min(us): 10505, Max(us): 41090, 99th(us): 35000, 99.9th(us): 40000, 99.99th(us): 41000
...
```

> Normally, this workload above will run several minutes, you will have enough time to check and manipulate the cluster.

Reviewer: ```suggestion
{{ <info> }}
This workload above will run several minutes, you will have enough time to simulate a node failure described as follows.
{{ </info> }}
```

## Step 5. Check the workload

1. Go to the [Grafana](https://grafana.com) page at [http://127.0.0.1:3000](http://127.0.0.1:3000)

2. Login with default username `admin` and password `admin`

3. Go to dashboard **playground-tikv-summary**, the OPS information is in panel **gRPC message count** in row **gRPC**.

{{< figure
    src="/img/docs/check-ops.png"
    width="80"
    number="1" >}}

4. By default, TiKV replicates all data 3 times and balances it across all stores. To see this balance, go to page **playground-overview** and check the region count across all nodes. In this example, we load a small amount of data, thus only one region is presented:

{{< figure
    src="/img/docs/fault-tolerance-region-count.png"
    width="80"
    number="1" >}}

## Step 6. Simulate a single node failure

> To understand fault tolerance in TiKV, it's important to review a few concepts from the [architecture](https://github.com/tikv/tikv#tikv-software-stack).

Reviewer: links to https://tikv.org/docs/5.1/reference/architecture/overview/

Author: Noted.

| Concept        |                                                   Description                                                    |
| -------------- | :--------------------------------------------------------------------------------------------------------------: |
| **Raft Group** |                  Each replica of a region is called Peer. All of such peers form a raft group.                   |
| **Leader**     | In every raft group, there is a unique role called leader, are responsible for read/write requests from clients. |

Notice that all read/write operations are handled by the leader of the region group, we are going to stop the only one leader in the cluster, and check the load continuity and cluster health.

1. Go to Grafana dashboard **playground-overview**, the leader distribution is in panel **leader** in row **TiKV**.

2. In the example, the local process that open port `20180` hold only one leader in the cluster. Use the following command to stop the process.
    ```sh
    kill -STOP $(lsof -i:20180 | grep tikv | head -n1 | awk '{print $2}')
    ```

## Step 7. Check load continuity and cluster health

> 1. Check the leader distribution again, you will find the leader is moving another store.

Reviewer: ```suggestion
1. Check the leader distribution again, you will find the leader is moving to another store.
```

{{< figure
    src="/img/docs/fault-tolerance-leader-recover.png"
    width="80"
    number="1" >}}

> 2. Check the gRPC OPS, you will find there is a small duration that the TiKV is unavailable because the leader was down. However, the workload is back online as soon as the [election](https://raft.github.io/raft.pdf) is completed.

Reviewer: ```suggestion
2. Check the gRPC OPS, you will find there is a small duration that the TiKV is unavailable because the leader was down. However, the workload is back online as soon as the election is completed.
```

Author: Because there is no such reference in the docs. I suggest keeping this link due to not all people know about the concept of `election`.

{{< figure
    src="/img/docs/fault-tolerance-ops.png"
    width="80"
    number="1" >}}

## Step 8. Prepare for two simultaneous node failures

> At this point, the cluster has recovered and is ready to handle another failure. In the example above, we stop the leader of the cluster which result in 5 stores are alive. Then, a new leader is presented after a while. We are going to increase the region replicas of TiKV to 5, stop 2 non-leader nodes simultaneously and check the cluster status.

Reviewer: ```suggestion
At this point, the cluster has recovered. In the example above, we stop the leader of the cluster which result in 5 stores are alive. Then, a new leader is presented after a while. We are going to increase the region replicas of TiKV to 5, stop 2 non-leader nodes simultaneously and check the cluster status.
```

---
title: Fault Tolerance and Recovery
description: Learn how TiKV recovers from failures
menu:
    "5.1":
        parent: Features
        weight: 2
---

This page walks you through a simple demonstration of how TiKV recovers from failures and continues to provide services during failures.

> 1. Starting a 6-node local cluster.

Reviewer 2: ```suggestion
1. Start a cluster with 6 TiKV nodes.
```

Author: The same mean to me.

Author: Referring https://www.cockroachlabs.com/docs/v21.1/demo-fault-tolerance-and-recovery.html#step-1-start-a-6-node-cluster

Reviewer 2: Updated suggestion

> 2. Run a sample workload via [go-ycsb](https://github.com/pingcap/go-ycsb), terminate a node to simulate a failure, and see how the cluster continues uninterrupted.

Reviewer 2: ```suggestion
2. Run a sample workload via [go-ycsb](https://github.com/pingcap/go-ycsb), then kill a node to simulate a failure, and see how the cluster seamlessly continues.
```

Author: Ditto.

Reviewer 2: This is also for uniformity. You used the word `kill` in the following step-by-step instruction, so it'll be better to use the same word for the same action. Besides, `seamlessly` sounds more stable than `uninterrupted` which is like 'TiKV will not shutdown but the performance may drop'.

Author: I am sorry, I can't see the difference. :cry:

Reviewer 2: kill vs terminate? no difference?

Author: Done.

3. Leave that node offline for long enough to watch the cluster repair itself by re-replicating missing data to other nodes.

> 4. Prepare the cluster for 2 simultaneous node failures by increasing to 5-way replication, then take two nodes offline at the same time, and again see how the cluster continues uninterrupted.

Reviewer 2: ```suggestion
4. Increase to 5 replications, then simultaneously take two nodes offline, and again see how the cluster recovers from failure without interrupting online service .
```

Author: Ditto.

Reviewer 2: This made the sentence more clear by reducing unnecessary long words and put the action you want to do to the very beginning.

Author: Accepting the former suggestion. 
```suggestion
4. Increase to 5 replications, then simultaneously take two nodes offline, and again see how the cluster continues uninterrupted.
```

## Prerequisites

1. Install [TiUP](https://github.com/pingcap/tiup) version **v1.5.2** or above as described in [TiKV in 5 Minutes](../../tikv-in-5-minutes)
2. Install [client-py](https://github.com/tikv/client-py) to interact with the TiKV cluster.

> 

Reviewer 2: ```suggestion
```

Author: Why remove this empty line?

Reviewer 2: Because there are two blank lines.

## Step 4. Run a sample workload.

Go to the source directory of `go-ycsb`, use the following command to run the `workloada` from YCSB benchmark, simulating multiple client connections, each performing mixed 50% read and 50% write operations.

```sh
./bin/go-ycsb run tikv -P workloads/workloada -p tikv.pd="127.0.0.1:2379" -p tikv.type="raw" -p tikv.conncount=16 -p threadcount=16 -p recordcount=10000 -p operationcount=1000000
```

You'll see per-operation statistics print to standard output every second.

```
...
READ   - Takes(s): 10.0, Count: 7948, OPS: 796.2, Avg(us): 395, Min(us): 72, Max(us): 20545, 99th(us): 2000, 99.9th(us): 19000, 99.99th(us): 21000
UPDATE - Takes(s): 10.0, Count: 7945, OPS: 796.8, Avg(us): 19690, Min(us): 11589, Max(us): 40429, 99th(us): 34000, 99.9th(us): 41000, 99.99th(us): 41000
READ   - Takes(s): 20.0, Count: 15858, OPS: 793.6, Avg(us): 380, Min(us): 68, Max(us): 20545, 99th(us): 2000, 99.9th(us): 3000, 99.99th(us): 21000
UPDATE - Takes(s): 20.0, Count: 15799, OPS: 791.1, Avg(us): 19834, Min(us): 10505, Max(us): 41090, 99th(us): 35000, 99.9th(us): 40000, 99.99th(us): 41000
...
```

> {{ <info> }}

Reviewer: ```suggestion
{{< info >}}
```

This workload above will run several minutes, you will have enough time to simulate a node failure described as follows.

> {{ </info> }}

Reviewer: ```suggestion
{{< /info >}}
```

Author: Done.

{{< figure
    src="/img/docs/check-ops.png"
    width="80"
    number="1" >}}

4. By default, TiKV replicates all data 3 times and balances it across all stores. To see this balance, go to page **playground-overview** and check the region count across all nodes. In this example, we load a small amount of data, thus only one region is presented:

{{< figure
    src="/img/docs/fault-tolerance-region-count.png"
    width="80"
    number="1" >}}

## Step 6. Simulate a single node failure

To understand fault tolerance in TiKV, it's important to review a few concepts from the [architecture](https://tikv.org/docs/5.1/reference/architecture/overview).
| Concept        |                                                   Description                                                    |
| -------------- | :--------------------------------------------------------------------------------------------------------------: |
| **Raft Group** |                  Each replica of a region is called Peer. All of such peers form a raft group.                   |
| **Leader**     | In every raft group, there is a unique role called leader, are responsible for read/write requests from clients. |

> 

Reviewer 2: I think this is unnecessary because of the link to the doc.

Author: This page intends to be self-explained. Just like the chapter: `replication and rebalance`

Reviewer 2: I think it's very very strange to me that you list the important definitions in between some messy steps. And because these definitions are the root concept of the topic of this page, you can safely assume that they are well known. This is not similar to `replication and rebalance`: you list the definition on the very top of the page; that topic is exactly talking about Raft and Regions; there are many other related concepts (while in here, you picked only two of the whole list).

Author: You can't assume the reader knows every concept. Any page of the docs tends to be self-explained, just like CRDB's does.

Reviewer 2: In case of that, you can put a link for the `Region` and `Raft`. Putting a fragile concept list is very distracting and confusing.

Author: Hmmm. The concept about Region and Raft has only pictures on page `Overview`. I insist that the any page of the docs should be self-explained.
